cpp/include/raft/linalg/add.cuh

```diff
@@ -72,24 +74,150 @@ void add(OutT* out, const InT* in1, const InT* in2, IdxType len, cudaStream_t st
 
 /** Substract single value pointed by singleScalarDev parameter in device memory from inDev[i] and
  * write result to outDev[i]
- * @tparam math_t data-type upon which the math operation will be performed
+ * @tparam InT     input data-type. Also the data-type upon which the math ops
+ *                 will be performed
+ * @tparam OutT    output data-type
  * @tparam IdxType Integer type used to for addressing
  * @param outDev the output buffer
  * @param inDev the input buffer
  * @param singleScalarDev pointer to the scalar located in device memory
  * @param len number of elements in the input and output buffer
  * @param stream cuda stream
  */
-template <typename math_t, typename IdxType = int>
-void addDevScalar(math_t* outDev,
-                  const math_t* inDev,
-                  const math_t* singleScalarDev,
-                  IdxType len,
-                  cudaStream_t stream)
+template <typename InT, typename OutT = InT, typename IdxType = int>
+void addDevScalar(
+  OutT* outDev, const InT* inDev, const InT* singleScalarDev, IdxType len, cudaStream_t stream)
 {
   detail::addDevScalar(outDev, inDev, singleScalarDev, len, stream);
 }
 
+/**
+ * @defgroup add Addition Arithmetic
+ * @{
+ */
+
+/**
+ * @brief Elementwise add operation on the input buffers
+ * @tparam InType    Input Type raft::device_mdspan
+ * @tparam OutType   Output Type raft::device_mdspan
+ * @param handle raft::handle_t
+ * @param out    Output
+ * @param in1    First Input
+ * @param in2    Second Input
+ */
+template <typename InType,
+          typename OutType = InType,
+          typename         = raft::enable_if_device_mdspan<OutType, InType>>
+void add(const raft::handle_t& handle, OutType out, const InType in1, const InType in2)
+{
+  using in_element_t  = typename InType::element_type;
+  using out_element_t = typename OutType::element_type;
+
+  RAFT_EXPECTS(out.is_exhaustive(), "Output must be contiguous");
+  RAFT_EXPECTS(in1.is_exhaustive(), "Input 1 must be contiguous");
+  RAFT_EXPECTS(in2.is_exhaustive(), "Input 2 must be contiguous");
+  RAFT_EXPECTS(out.size() == in1.size() && in1.size() == in2.size(),
+               "Size mismatch between Output and Inputs");
+
+  if (out.size() <= std::numeric_limits<std::uint32_t>::max()) {
```

Comment:
In point 2, what happens in extreme cases? Consider `index_type=uint32_t` with extents `{2^32, 2}`. In this case, will `required_span_size()` by representable by `index_type` or will it cause an overflow?

Reply:
@divyegala `required_span_size()` is _not_ representable by `index_type` in this case.  For `layout_left` and `layout_right`, `required_span_size()` and `size()` are the same mathematically.  The only difference is the return type (`index_type` resp. `size_t`).  For `layout_stride`, though, `required_span_size()` can be greater than the `size()`.  For other layouts (e.g., the "matrix of a single value" layout that maps all multidimensional indices to the offset zero), `required_span_size()` can be less than `size()`. 

Note that while it's UB for users to violate preconditions, implementations aren't required to _check_ preconditions.  The reference implementation of `layout_left` does _not_ currently check preconditions, as [you can see here](https://github.com/kokkos/mdspan/blob/d58ddaaa1869aa5ad8469c3b02abe025efab52a4/include/experimental/__p0009_bits/layout_left.hpp#L117), for instance.  This means two things.

1. If someone gives you a `layout_{left,right,stride}::mapping` instance (e.g., in an `mdspan`), then you can assume that the precondition is satisfied.

2. If you are _constructing_ a `layout_{left,right,stride}::mapping` instance (e.g., by constructing an `mdspan` with a pointer and extents), then you are responsible for ensuring that the precondition is satisfied.

Reply:
@divyegala wrote:
> In this case, will `required_span_size()` by representable by `index_type` or will it cause an overflow?

Those are two separate questions, actually! : - ) 

1. `required_span_size()` is _not_ representable by `index_type` in this case.
2. Giving this `extents` object to `layout_{left,right,stride}::mapping`'s constructor violates the constructor's precondition.  It could overflow, or it could open a portal to the Awesome Dimension and let loose a swarm of nasal demons who search out precondition violators and boop them gently on the nose.

Reply:
@mhoemmen thanks for the explanations! How do we really represent such edge cases and safely obtain the product of the extents? Sounds like `size()` is the safe way to obtain the product without violating any pre-conditions since it's representable by `size_t`?

Reply:
@divyegala Gladly! : - )

> How do we really represent such edge cases and safely obtain the product of the extents? 

By the time the user has created a layout mapping, it's already too late.  What I mean by that is that if `required_span_size()` doesn't fit `index_type`, then the user will likely get the wrong answer when they try to index into the mdspan.

In what follows in my comment, I'll distinguish between "the Preconditions in the spec" and "what the reference implementation does."  The reference implementation currently does _not_ check this precondition in the layout mapping.  This means that it's possible for users to construct extents for which the mapping's `required_span_size()` can overflow.

We can prevent this by wrapping mdspan creation to check the extents object for potential overflow, before it goes into a layout mapping's constructor.  It's not UB to construct, e.g., `dextents<uint16_t, 2>(` 2^{15} `,` 2^{15} `)`.  We just need to intercept that naughty `extents` value before it goes into a layout mapping's constructor.  Otherwise, the layout mapping has the freedom to do whatever it likes, including calling `abort()`.

Our mdarray implementation's conversion to mdspan can also check, but again, we're probably better off making the wrapper explicit and not part of the mdarray proposal.  WG21 likes Preconditions and wants violating them to be UB.  If we want some specified behavior (e.g., throwing a particular exception, or calling `terminate()` after printing a helpful error message), then we'll have to implement that ourselves.